File: android/src/main/kotlin/com/usercentrics/sdk/flutter/serializer/UserDecisionSerializer.kt
package com.usercentrics.sdk.flutter.serializer

import com.usercentrics.sdk.UserDecision
import com.usercentrics.sdk.services.tcf.interfaces.TCFUserDecisionOnPurpose
import com.usercentrics.sdk.services.tcf.interfaces.TCFUserDecisionOnSpecialFeature
import com.usercentrics.sdk.services.tcf.interfaces.TCFUserDecisionOnVendor
import com.usercentrics.sdk.services.tcf.interfaces.TCFUserDecisions

internal fun Any?.deserializeUserDecision(): UserDecision {
    this as Map<*, *>
    return UserDecision(
        serviceId = this["serviceId"] as String,
        consent = this["consent"] as Boolean,
    )
}

internal fun Any?.deserializeTCFUserDecisions(): TCFUserDecisions {
    this as Map<*, *>
    return TCFUserDecisions(
        purposes = (this["purposes"] as List<*>).map { it.deserializeTCFUserDecisionOnPurpose() },
        specialFeatures = (this["specialFeatures"] as List<*>).map { it.deserializeTCFUserDecisionOnSpecialFeature() },
        vendors = (this["vendors"] as List<*>).map { it.deserializeTCFUserDecisionOnVendor() },
    )
}

private fun Any?.deserializeTCFUserDecisionOnPurpose(): TCFUserDecisionOnPurpose {
    this as Map<*, *>
    return TCFUserDecisionOnPurpose(
        id = this["id"] as Int,
        consent = this["consent"] as Boolean?,
        legitimateInterestConsent = this["legitimateInterestConsent"] as Boolean?,
    )
}

private fun Any?.deserializeTCFUserDecisionOnSpecialFeature(): TCFUserDecisionOnSpecialFeature {
    this as Map<*, *>
    return TCFUserDecisionOnSpecialFeature(
        id = this["id"] as Int,
        consent = this["consent"] as Boolean?,
    )
}

private fun Any?.deserializeTCFUserDecisionOnVendor(): TCFUserDecisionOnVendor {
    this as Map<*, *>
    return TCFUserDecisionOnVendor(
        id = this["id"] as Int,
        consent = this["consent"] as Boolean?,
        legitimateInterestConsent = this["legitimateInterestConsent"] as Boolean?,
    )
}

File: lib/src/internal/serializer/user_decision_serializer.dart
import 'package:usercentrics_sdk/src/model/user_decision.dart';

class UserDecisionSerializer {
  static dynamic serialize(UserDecision value) => {
        'serviceId': value.serviceId,
        'consent': value.consent,
      };
}

class TCFUserDecisionsSerializer {
  static dynamic serialize(TCFUserDecisions value) => {
        'purposes': value.purposes.map((e) => _serializePurpose(e)).toList(),
        'specialFeatures':
            value.specialFeatures.map((e) => _serializeFeature(e)).toList(),
        'vendors': value.vendors.map((e) => _serializeVendor(e)).toList(),
      };

  static dynamic _serializePurpose(TCFUserDecisionOnPurpose value) => {
        'id': value.id,
        'consent': value.consent,
        'legitimateInterestConsent': value.legitimateInterestConsent,
      };

  static dynamic _serializeFeature(TCFUserDecisionOnSpecialFeature value) => {
        'id': value.id,
        'consent': value.consent,
      };

  static dynamic _serializeVendor(TCFUserDecisionOnVendor value) => {
        'id': value.id,
        'consent': value.consent,
        'legitimateInterestConsent': value.legitimateInterestConsent,
      };
}
